rust_snuba/src/processors/generic_metrics.rs

```diff
@@ -0,0 +1,1173 @@
+    tags_indexed_value: Vec<u64>,
+    #[serde(rename = "tags.raw_value")]
+    tags_raw_value: Vec<String>,
+    #[serde(default)]
+    set_values: Vec<u64>,
+    #[serde(default)]
+    count_value: f64,
+    #[serde(default)]
+    distribution_values: Vec<f64>,
+    metric_type: String,
+    materialization_version: u8,
+    timeseries_id: u32,
+    partition: u16,
+    offset: u64,
+    granularities: Vec<u8>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    decasecond_retention_days: Option<u8>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    min_retention_days: Option<u8>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    hr_retention_days: Option<u8>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    day_retention_days: Option<u8>,
+}
+
+impl TryFrom<FromGenericMetricsMessage> for CountersRawRow {
+    type Error = anyhow::Error;
+
+    fn try_from(from: FromGenericMetricsMessage) -> anyhow::Result<CountersRawRow> {
+        if from.version != 2 {
+            // TODO: No checks happen for version even in current python processor. Should
+            // we be doing something here?
+            tracing::warn!("Unsupported version: {}", from.version);
+        }
+
+        if from.type_ != "c" {
+            return Err(anyhow::anyhow!("Unsupported metric type: {}", from.type_));
+        }
+
+        let timeseries_id =
+            generate_timeseries_id(from.org_id, from.project_id, from.metric_id, &from.tags);
+
+        let (tag_keys, tag_values): (Vec<_>, Vec<_>) = from.tags.into_iter().unzip();
+
+        let mut granularities = vec![
+            GRANULARITY_ONE_MINUTE,
+            GRANULARITY_ONE_HOUR,
+            GRANULARITY_ONE_DAY,
+        ];
+
+        if from.aggregation_option.unwrap_or_default() == "ten_second" {
+            granularities.push(GRANULARITY_TEN_SECONDS);
+        }
+
+        let count_value = match from.value {
+            MetricValue::Counter(value) => value,
+            _ => {
+                return Err(anyhow::anyhow!(
+                    "Unsupported values provided for counter metric type"
+                ))
+            }
+        };
+
+        Ok(Self {
+            use_case_id: from.use_case_id,
+            org_id: from.org_id,
+            project_id: from.project_id,
+            metric_type: "counter".to_string(),
+            metric_id: from.metric_id,
+            timestamp: from.timestamp,
+            retention_days: from.retention_days,
+            tags_key: tag_keys.iter().map(|k| k.parse::<u64>().unwrap()).collect(),
+            tags_indexed_value: vec![0; tag_keys.len()],
+            tags_raw_value: tag_values,
+            count_value,
+            materialization_version: 2,
+            timeseries_id,
+            granularities,
+            min_retention_days: None,
+            decasecond_retention_days: None,
+            hr_retention_days: None,
+            day_retention_days: None,
```

Comment:
What's all this retention days stuff? I don't see similar in the python version of the processor

Reply:
The clickhouse tables basically have a way to customize retention of different granularities. There are defaults in some of the fields. But those can be overriden. Example here https://github.com/getsentry/snuba/blob/master/snuba/datasets/metrics_messages.py#L106-L112.